lib/common/common/src/cpu.rs

```diff
@@ -1,7 +1,196 @@
+
+    /// Try to acquire CPU permit for optimization task from global CPU budget.
+    ///
+    /// The given `desired_cpus` is not exact, but rather a hint on what we'd like to acquire.
+    /// - it will prefer to acquire the maximum number of CPUs
+    /// - it will never be higher than the total CPU budget
+    /// - it will never be lower than `min_permits(desired_cpus)`
+    pub fn try_acquire(&self, desired_cpus: usize) -> Option<CpuPermit> {
+        // Determine what number of CPUs to acquire based on available budget
+        let min_required = self.min_permits(desired_cpus) as u32;
+        let num_cpus = self.semaphore.available_permits().min(desired_cpus) as u32;
+        if desired_cpus == 0 || num_cpus < min_required {
+            return None;
+        }
+
+        // Try to acquire selected number of CPUs
+        let result = Semaphore::try_acquire_many_owned(self.semaphore.clone(), num_cpus);
+        let permit = match result {
+            Ok(permit) => permit,
+            Err(TryAcquireError::NoPermits) => return None,
+            Err(TryAcquireError::Closed) => unreachable!("Cannot acquire CPU permit because CPU budget semaphore is closed, this should never happen"),
+        };
+
+        Some(CpuPermit::new(num_cpus, permit))
+    }
+
+    /// Check if there is enough CPU budget available for the given `desired_cpus`.
+    ///
+    /// This checks for the minimum number of required permits based on the given desired CPUs,
+    /// based on `min_permits`. To check for an exact number, use `has_budget_exact` instead.
+    ///
+    /// A desired CPU count of `0` will always return `true`.
+    pub fn has_budget(&self, desired_cpus: usize) -> bool {
+        self.has_budget_exact(self.min_permits(desired_cpus))
+    }
+
+    /// Check if there are at least `budget` available CPUs in this budget.
+    ///
+    /// A budget of `0` will always return `true`.
+    fn has_budget_exact(&self, budget: usize) -> bool {
+        self.semaphore.available_permits() >= budget
+    }
+
+    /// Wait until we have CPU budget available for the given number of desired CPUs.
+    ///
+    /// Waits for at least the minimum number of permits based on the given desired CPUs. For
+    /// example, if `desired_cpus` is 8, this will wait for at least 4 to be available. See
+    /// [`Self::min_permits`].
+    ///
+    /// - `1` to wait for any CPU budget to be available.
+    /// - `0` will always return immediately.
+    ///
+    /// Uses an exponential backoff strategy up to 10 seconds to avoid busy waiting.
+    pub async fn wait_until_budget(&self, desired_cpus: usize) {
+        let min_required = self.min_permits(desired_cpus);
+        if self.has_budget_exact(min_required) {
+            return;
+        }
+
+        // Wait for CPU budget to be available with exponential backoff
+        // TODO: find better way, don't busy wait
+        let mut delay = Duration::from_micros(100);
+        while !self.has_budget_exact(min_required) {
```

Comment:
Can we have an upper bound on the wait time here instead of waiting forever?

Reply:
Is there a good reason for setting an upper bound in this case? I don't really see a way for this getting stuck forever.

Reply:
>  I don't really see a way for this getting stuck forever. It it would get stuck, it wouldn't be a huge problem.

Is it because the number of waiting tasks is at most equal to the number of optimizers running?

Reply:
The number of waiting tasks never exceeds 1 per shard. That is because we only spawn a 'wait task' once per shard if we run out of CPU budget.

Waiting this way is fair, meaning that there would be no waiter having an unfair advantage, always going first.
Though, a shard with a lot of ongoing updates may claim CPU budget before our wait function notices, but I don't see that as a problem.

CPU budget is eventually always returned once optimizations complete because it's handled in on-drop.

Now, if an optimization would get stuck somehow CPU budget is not returned. We could timeout in this wait function, but it wouldn't make much sense. If we'd do that, we still wouldn't be able to claim CPU budget afterwards.

Maybe it's important to add this: every time an upsertion happens, we try to claim CPU budget again. We're never stuck at this wait function, it's only spawned once on the side to trigger optimizations one more time. That is so we ensure we also optimize if we don't have upsertions.

Here it is in code. If we do any upsertion, we always go over [this](https://github.com/qdrant/qdrant/blob/08930fefa940729064f649c1f7ec0bef53daad66/lib/collection/src/update_handler.rs#L469) if-statement. If we have budget available at that time, we continue and try to optimize. If we don't, we make sure we spawn a separate task [here](https://github.com/qdrant/qdrant/blob/08930fefa940729064f649c1f7ec0bef53daad66/lib/collection/src/update_handler.rs#L473C25-L479C26) (once) with the above wait function, it triggers this loop again when CPU budget is available. While that task is ongoing, all new upsertions will try to pass this if-statement again.

Sorry for the wall of text, but I hope it clarifies :smile: 

Based on that, do you still think this can be an issue?

Reply:
Thank you very much for the wall of text, it starts clearing up the fog :face_in_clouds: 

I think my only remaining concerns is the number of notification tasks.
In case of a very heavy upsertion stream, are we going to spawn a lot of notification tasks if an HNSW indexing is currently running?

Reply:
> In case of a very heavy upsertion stream, are we going to spawn a lot of notification tasks if an HNSW indexing is currently running?

No, a maximum of 1 per shard. And there is no concurrency here, since each shard has a single optimization worker.

So if you'd have 1000 shards, then you could reach 1000 notification tasks. But that would be an unreasonable number anyway.